plugins/cassandra/cassandraembed.hpp

```diff
@@ -165,48 +191,98 @@ class CassandraStatement : public CInterface
+    void bindBool(unsigned idx, cass_bool_t value)
     {
-        //DBGLOG("bind %d %s", idx, value);
-        check(cass_statement_bind_string(statement, idx, value));
+        if (query.length())
+            traceBind(idx, "%s", value ? "true" : "false");
+        check(cass_statement_bind_bool(statement, idx, value));
     }
-    inline void bindString_n(unsigned idx, const char *value, unsigned len)
+    void bindInt32(unsigned idx, __int32 value)
     {
-        //DBGLOG("bind %d %.*s", idx, len, value);
-        check(cass_statement_bind_string_n(statement, idx, value, len));
+        if (query.length())
+            traceBind(idx, "%d", value);
+        check(cass_statement_bind_int32(statement, idx, value));
     }
-private:
-    CassandraStatement(const CassandraStatement &);
-    CassStatement *statement;
-};
-
-class CassandraPrepared : public CInterfaceOf<IInterface>
-{
-public:
-    inline CassandraPrepared(const CassPrepared *_prepared) : prepared(_prepared)
+    void bindInt64(unsigned idx, __int64 value)
     {
+        if (query.length())
+            traceBind(idx, "%" I64F "d", value);
+        check(cass_statement_bind_int64(statement, idx, value));
     }
-    inline ~CassandraPrepared()
+    void bindString(unsigned idx, const char *value)
     {
-        if (prepared)
-            cass_prepared_free(prepared);
+        if (query.length())
+            traceBind(idx, "'%s'", value);
+        check(cass_statement_bind_string(statement, idx, value));
     }
-    inline operator const CassPrepared *() const
+    void bindString_n(unsigned idx, const char *value, unsigned len)
     {
-        return prepared;
+        if (strlen(value)<len)
+        {
+            if (query.length())
+                traceBind(idx, "'%s'", value);
+            check(cass_statement_bind_string(statement, idx, value));
+        }
+        else
+        {
+            if (query.length())
+                traceBind(idx, "'%.*s'", len, value);
+            check(cass_statement_bind_string_n(statement, idx, value, len));
+        }
+    }
+    void bindBytes(unsigned idx, const cass_byte_t *value, unsigned len)
+    {
+        if (query.length())
+            traceBind(idx, "(bytes)");
+        check(cass_statement_bind_bytes(statement, idx, value, len));
+    }
+    void bindCollection(unsigned idx, const CassCollection *value)
+    {
+        if (query.length())
+            traceBind(idx, "(collection)");
+        check(cass_statement_bind_collection(statement, idx, value));
     }
 private:
-    CassandraPrepared(const CassandraPrepared &);
-    const CassPrepared *prepared;
+    void traceBind(unsigned idx, const char *format, ...)
+    {
+        assert(query.length());
+        StringBuffer bound;
+        va_list args;
+        va_start(args, format);
+        bound.valist_appendf(format, args);
+        va_end(args);
+        const char *pos = query;
+        do
+        {
+            pos = strchr(pos, '?');
+            assert(pos);
```

Comment:
It's only going to appear in debug builds (well, hopefully it's never going to appear at all), where I think an assert is quite useful...